src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,192 @@
+                if (x.item == null) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> x = first; x != null; x = x.next) {
+                if (isEquals(object, x.item)) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> prev;
+
+        Node(Node<T> prev, T element, Node<T> next) {
+            this.item = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
+    private Node<T> addFirst(T value) {
+        Node<T> newNode = new Node<>(null, value, null);
+        first = newNode;
+        last = newNode;
+        return newNode;
+    }
+
+    private Node<T> addLast(T value) {
+        Node<T> newNode = new Node<>(last, value, null);
+        last.next = newNode;
+        last = newNode;
+        return newNode;
+    }
+
+    private void checkPositionIndex(int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("Incorrect index: " + index);
+        }
+    }
+
+    private void checkIndexException(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of range: " + index);
+        }
+    }
+
+    Node<T> node(int index) {
+        Node<T> x;
+        if (index < (size >> 1)) {
+            x = first;
+            for (int i = 0; i < index; i++) {
+                x = x.next;
+            }
+        } else {
+            x = last;
+            for (int i = size - 1; i > index; i--) {
+                x = x.prev;
+            }
+        }
+        return x;
+    }
+
+    T unlink(Node<T> node) {
+        final T element = node.item;
+        Node<T> next = node.next;
+        Node<T> prev = node.prev;
+        if (prev == null) {
+            first = next;
+        } else {
+            prev.next = next;
+            node.prev = null;
+        }
+        if (next == null) {
+            last = prev;
+        } else {
+            next.prev = prev;
+            node.next = null;
+        }
+        node.item = null;
+        size--;
+        return element;
+    }
+
+    private boolean isEquals(T a, T b) {
+        return a == b || a != null && a.equals(b);
+    }
+
+    void linkLast(T e) {
+        final Node<T> l = last;
+        final Node<T> newNode = new Node<>(l, e, null);
+        last = newNode;
+        if (l == null) {
+            first = newNode;
+        } else {
+            l.next = newNode;
+        }
+        size++;
+    }
+
+    void linkFirst(T e, Node<T> succ) {
```

Comment:
What if `succ`? Let's rename it somehow.

```suggestion
    void linkFirst(T element, Node<T> node {
```